ui/coordinateTrainer/src/ctrl.ts

```diff
@@ -0,0 +1,176 @@
+import { sparkline } from '@fnando/sparkline';
+import * as xhr from 'common/xhr';
+import throttle from 'common/throttle';
+import { Api as CgApi } from 'chessground/api';
+import { ColorChoice, CoordinateTrainerConfig, Redraw } from './interfaces';
+
+const orientationFromColorChoice = (colorChoice: ColorChoice): Color =>
+  (colorChoice === 'random' ? ['white', 'black'][Math.round(Math.random())] : colorChoice) as Color;
+
+const newKey = (oldKey: Key | ''): Key => {
+  // disallow the previous coordinate's row or file from being selected
+  const files = 'abcdefgh'.replace(oldKey[0], '');
+  const rows = '12345678'.replace(oldKey[1], '');
+  return (files[Math.floor(Math.random() * files.length)] + rows[Math.floor(Math.random() * files.length)]) as Key;
+};
+
+export const DURATION = 30 * 1000;
+const TICK_DELAY = 50;
+
+export default class CoordinateTrainerCtrl {
+  config: CoordinateTrainerConfig;
+  colorChoice: ColorChoice;
+  orientation: Color;
+  isAuth: boolean;
+  playing = false;
+  hasPlayed = false;
+  trans: Trans;
+  chessground: CgApi | undefined;
+  redraw: Redraw;
+  score = 0;
+  currentKey: Key | '' = 'a1';
+  nextKey: Key | '' = newKey('a1');
+  timeLeft = DURATION;
+  timeAtStart: Date;
+  wrongTimeout: number;
+  wrong: boolean;
+  zen: boolean;
+
+  constructor(config: CoordinateTrainerConfig, redraw: Redraw) {
+    this.config = config;
+    this.colorChoice = config.colorPref || 'random';
+    this.orientation = orientationFromColorChoice(this.colorChoice);
+
+    this.isAuth = document.body.hasAttribute('data-user');
+    this.trans = lichess.trans(this.config.i18n);
+    this.redraw = redraw;
+
+    const setZen = throttle(1000, zen =>
+      xhr.text('/pref/zen', {
+        method: 'post',
+        body: xhr.form({ zen: zen ? 1 : 0 }),
+      })
+    );
+
+    lichess.pubsub.on('zen', () => {
+      const zen = $('body').toggleClass('zen').hasClass('zen');
+      window.dispatchEvent(new Event('resize'));
+      setZen(zen);
+      requestAnimationFrame(this.updateCharts);
+    });
+
+    window.Mousetrap.bind('z', () => lichess.pubsub.emit('zen'));
+
+    $('#zentog').on('click', () => lichess.pubsub.emit('zen'));
+
+    window.addEventListener('resize', () => requestAnimationFrame(this.updateCharts), true);
```

Comment:
We update the charts on window resize to fix this bug:
![image](https://user-images.githubusercontent.com/12655228/157112814-ee0926f9-9d15-44af-ae25-3c54b155434c.png)